File: SimplifAI/DocumentVerificationUI/src/app/services/security.service.ts
import { Injectable } from '@angular/core';
import { HttpRequest } from '@angular/common/http';

export interface SecurityValidationResult {
  isValid: boolean;
  errors: string[];
  warnings: string[];
}

export interface InputSanitizationOptions {
  maxLength?: number;
  allowHtml?: boolean;
  allowSpecialChars?: boolean;
  trimWhitespace?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class SecurityService {

  private readonly suspiciousPatterns = [
    // XSS patterns
    /<script[^>]*>.*?<\/script>/gi,
    /javascript:/gi,
    /vbscript:/gi,
    /on\w+\s*=/gi,
    /<iframe[^>]*>/gi,
    /<object[^>]*>/gi,
    /<embed[^>]*>/gi,
    
    // SQL injection patterns
    /(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|UNION)\b)/gi,
    /(\b(OR|AND)\s+\d+\s*=\s*\d+)/gi,
    /(--|#|\/\*|\*\/)/g,
    
    // Directory traversal
    /\.\.\//g,
    /\.\.\\/g,
    
    // Command injection
    /(\||&|;|\$\(|`)/g
  ];

  private readonly dangerousFileExtensions = [
    '.exe', '.bat', '.cmd', '.scr', '.com', '.pif', '.vbs', '.js', '.jar',
    '.app', '.deb', '.pkg', '.dmg', '.msi', '.run', '.sh', '.ps1', '.php',
    '.asp', '.aspx', '.jsp', '.py', '.rb', '.pl', '.cgi'
  ];

  constructor() {}

  /**
   * Sanitizes user input to prevent XSS and injection attacks
   */
  sanitizeInput(input: string, options: InputSanitizationOptions = {}): string {
    if (!input) return '';

    const {
      maxLength = 1000,
      allowHtml = false,
      allowSpecialChars = true,
      trimWhitespace = true
    } = options;

    let sanitized = input;

    // Trim whitespace if requested
    if (trimWhitespace) {
      sanitized = sanitized.trim();
    }

    // Remove HTML tags if not allowed
    if (!allowHtml) {
      sanitized = sanitized.replace(/<[^>]*>/g, '');
    }

    // Remove script content
    sanitized = sanitized.replace(/<script[^>]*>.*?<\/script>/gi, '');

    // Remove dangerous JavaScript events
    sanitized = sanitized.replace(/on\w+\s*=/gi, '');

    // Remove suspicious patterns
    this.suspiciousPatterns.forEach(pattern => {
      sanitized = sanitized.replace(pattern, '');
    });

    // Remove null bytes and control characters
    sanitized = sanitized.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');

    // Remove special characters if not allowed
    if (!allowSpecialChars) {
      sanitized = sanitized.replace(/[<>'"&]/g, '');
    }

    // Truncate to max length
    if (sanitized.length > maxLength) {
      sanitized = sanitized.substring(0, maxLength);
    }

    return sanitized;
  }

  /**
   * Validates input for security threats
   */
  validateInput(input: string): SecurityValidationResult {
    const result: SecurityValidationResult = {
      isValid: true,
      errors: [],
      warnings: []
    };

    if (!input) {
      return result;
    }

    // Check for suspicious patterns
    this.suspiciousPatterns.forEach(pattern => {
      if (pattern.test(input)) {
        result.isValid = false;
        result.errors.push('Input contains potentially dangerous content');
      }
    });

    // Check for excessively long input
    if (input.length > 10000) {
      result.isValid = false;
      result.errors.push('Input is too long');
    }

    // Check for null bytes
    if (input.includes('\0')) {
      result.isValid = false;
      result.errors.push('Input contains null bytes');
    }

    // Warn about special characters
    if (/[<>"'&]/.test(input)) {
      result.warnings.push('Input contains special characters that may be encoded');
    }

    return result;
  }

  /**
   * Validates file security before upload
   */
  validateFileUpload(file: File): SecurityValidationResult {
    const result: SecurityValidationResult = {
      isValid: true,
      errors: [],
      warnings: []
    };

    // Check file size (10MB limit)
    const maxSize = 10 * 1024 * 1024;
    if (file.size > maxSize) {
      result.isValid = false;
      result.errors.push(`File size (${this.formatFileSize(file.size)}) exceeds maximum allowed size (${this.formatFileSize(maxSize)})`);
    }

    // Check minimum file size
    if (file.size < 1024) {
      result.isValid = false;
      result.errors.push('File is too small and may be corrupted');
    }

    // Check filename
    if (!file.name || file.name.trim().length === 0) {
      result.isValid = false;
      result.errors.push('File must have a valid name');
    }

    // Check for dangerous file extensions
    const extension = this.getFileExtension(file.name).toLowerCase();
    if (this.dangerousFileExtensions.includes(extension)) {
      result.isValid = false;
      result.errors.push(`File extension '${extension}' is not allowed for security reasons`);
    }

    // Check for allowed extensions
    const allowedExtensions = ['.jpg', '.jpeg', '.png', '.pdf'];
    if (!allowedExtensions.includes(extension)) {
      result.isValid = false;
      result.errors.push(`File extension '${extension}' is not allowed`);
    }

    // Check MIME type
    const allowedMimeTypes = ['image/jpeg', 'image/png', 'application/pdf'];
    if (!allowedMimeTypes.includes(file.type.toLowerCase())) {
      result.isValid = false;
      result.errors.push(`File type '${file.type}' is not supported`);
    }

    // Check for suspicious filename patterns
    if (this.hasSuspiciousFilename(file.name)) {
      result.isValid = false;
      result.errors.push('Filename contains suspicious patterns');
    }

    // Check for double extensions
    const extensionCount = (file.name.match(/\./g) || []).length;
    if (extensionCount > 1) {
      result.warnings.push('File has multiple extensions - ensure this is legitimate');
    }

    // Warn about very large files
    if (file.size > 5 * 1024 * 1024) {
      result.warnings.push('Large file detected - may take longer to process');
    }

    return result;
  }

  /**
   * Validates URL for security
   */
  validateUrl(url: string): SecurityValidationResult {
    const result: SecurityValidationResult = {
      isValid: true,
      errors: [],
      warnings: []
    };

    if (!url) {
      return result;
    }

    try {
      const urlObj = new URL(url);
      
      // Check protocol
      if (!['http:', 'https:'].includes(urlObj.protocol)) {
        result.isValid = false;
        result.errors.push('URL must use HTTP or HTTPS protocol');
      }

      // Check for suspicious patterns in URL
      if (this.suspiciousPatterns.some(pattern => pattern.test(url))) {
        result.isValid = false;
        result.errors.push('URL contains suspicious patterns');
      }

      // Warn about non-HTTPS URLs
      if (urlObj.protocol === 'http:') {
        result.warnings.push('URL uses insecure HTTP protocol');
      }

    } catch (error) {
      result.isValid = false;
      result.errors.push('Invalid URL format');
    }

    return result;
  }

  /**
   * Generates a secure token for client-side use
   */
  generateSecureToken(): string {
    const array = new Uint8Array(32);
    crypto.getRandomValues(array);
    return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
  }

  /**
   * Validates a security token format
   */
  validateSecurityToken(token: string): boolean {
    if (!token || token.length < 32) {
      return false;
    }

    // Check for valid hexadecimal characters
    return /^[a-fA-F0-9]+$/.test(token);
  }

  /**
   * Adds security headers to HTTP requests
   */
  addSecurityHeaders(request: HttpRequest<any>): HttpRequest<any> {
    return request.clone({
      setHeaders: {
        'X-Requested-With': 'XMLHttpRequest',
        'X-Content-Type-Options': 'nosniff',
        'X-Frame-Options': 'DENY'
      }
    });
  }

  /**
   * Validates form data before submission
   */
  validateFormData(formData: any): SecurityValidationResult {
    const result: SecurityValidationResult = {
      isValid: true,
      errors: [],
      warnings: []
    };

    if (!formData) {
      return result;
    }

    // Recursively validate all form fields
    this.validateObjectFields(formData, result, '');

    return result;
  }

  /**
   * Checks if the current environment is secure (HTTPS)
   */
  isSecureEnvironment(): boolean {
    return window.location.protocol === 'https:' || 
           window.location.hostname === 'localhost' ||
           window.location.hostname === '127.0.0.1';
  }

  /**
   * Logs security events for monitoring
   */
  logSecurityEvent(eventType: string, details: string): void {
    console.warn(`Security Event: ${eventType} - ${details}`);
    
    // In production, you might want to send this to a logging service
    // this.http.post('/api/security/log', { eventType, details, timestamp: new Date() }).subscribe();
  }

  // Private helper methods

  private getFileExtension(filename: string): string {
    const lastDotIndex = filename.lastIndexOf('.');
    return lastDotIndex !== -1 ? filename.substring(lastDotIndex) : '';
  }

  private hasSuspiciousFilename(filename: string): boolean {
    // Check for directory traversal
    if (filename.includes('..') || filename.includes('\\') || filename.includes('/')) {
      return true;
    }

    // Check for null bytes
    if (filename.includes('\0')) {
      return true;
    }

    // Check for Windows reserved names
    const reservedNames = ['CON', 'PRN', 'AUX', 'NUL', 'COM1', 'COM2', 'COM3', 'COM4', 'COM5', 'COM6', 'COM7', 'COM8', 'COM9', 'LPT1', 'LPT2', 'LPT3', 'LPT4', 'LPT5', 'LPT6', 'LPT7', 'LPT8', 'LPT9'];
    const nameWithoutExtension = filename.split('.')[0].toUpperCase();
    if (reservedNames.includes(nameWithoutExtension)) {
      return true;
    }

    return false;
  }

  private validateObjectFields(obj: any, result: SecurityValidationResult, path: string): void {
    for (const [key, value] of Object.entries(obj)) {
      const currentPath = path ? `${path}.${key}` : key;
      
      if (typeof value === 'string') {
        const validation = this.validateInput(value);
        if (!validation.isValid) {
          result.isValid = false;
          validation.errors.forEach(error => {
            result.errors.push(`${currentPath}: ${error}`);
          });
        }
        validation.warnings.forEach(warning => {
          result.warnings.push(`${currentPath}: ${warning}`);
        });
      } else if (typeof value === 'object' && value !== null) {
        this.validateObjectFields(value, result, currentPath);
      }
    }
  }

  private formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
  }
}

File: SimplifAI/DocumentVerificationAPI/Services/SecurityService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace DocumentVerificationAPI.Services
{
    public class SecurityService : ISecurityService
    {
        private readonly ILogger<SecurityService> _logger;
        private readonly IConfiguration _configuration;
        private readonly string[] _allowedFileExtensions;
        private readonly string[] _allowedMimeTypes;
        private readonly long _maxFileSize;
        private readonly HashSet<string> _dangerousExtensions;
        private readonly HashSet<string> _suspiciousPatterns;

        public SecurityService(ILogger<SecurityService> logger, IConfiguration configuration)
        {
            _logger = logger;
            _configuration = configuration;
            
            _allowedFileExtensions = configuration.GetSection("Security:AllowedFileExtensions").Get<string[]>() ?? 
                new[] { ".jpg", ".jpeg", ".png", ".pdf" };
            _allowedMimeTypes = configuration.GetSection("Security:AllowedMimeTypes").Get<string[]>() ?? 
                new[] { "image/jpeg", "image/png", "application/pdf" };
            _maxFileSize = configuration.GetValue<long?>("Security:MaxFileSizeInBytes") ?? (10 * 1024 * 1024);

            _dangerousExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                ".exe", ".bat", ".cmd", ".scr", ".com", ".pif", ".vbs", ".js", ".jar",
                ".app", ".deb", ".pkg", ".dmg", ".msi", ".run", ".sh", ".ps1", ".php",
                ".asp", ".aspx", ".jsp", ".py", ".rb", ".pl", ".cgi"
            };

            _suspiciousPatterns = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "<script", "javascript:", "vbscript:", "onload=", "onerror=", "onclick=",
                "eval(", "exec(", "system(", "shell_exec", "passthru", "file_get_contents",
                "SELECT", "INSERT", "UPDATE", "DELETE", "DROP", "CREATE", "ALTER",
                "UNION", "OR 1=1", "' OR '1'='1", "admin'--", "'; DROP TABLE"
            };
        }

        public bool ValidateRequestSource(HttpRequest request)
        {
            try
            {
                // Check for required headers
                if (!request.Headers.ContainsKey("User-Agent"))
                {
                    LogSecurityEvent(SecurityEventType.SuspiciousInput, "Missing User-Agent header", request);
                    return false;
                }

                // Check for suspicious user agents
                var userAgent = request.Headers["User-Agent"].ToString();
                var suspiciousUserAgents = new[] { "sqlmap", "nikto", "nmap", "masscan", "curl", "wget" };
                
                if (suspiciousUserAgents.Any(ua => userAgent.Contains(ua, StringComparison.OrdinalIgnoreCase)))
                {
                    LogSecurityEvent(SecurityEventType.SuspiciousInput, $"Suspicious User-Agent: {userAgent}", request);
                    return false;
                }

                // Validate Content-Type for POST requests
                if (request.Method == "POST" && request.HasFormContentType)
                {
                    var contentType = request.ContentType?.ToLowerInvariant();
                    if (contentType != null && !contentType.StartsWith("multipart/form-data") && 
                        !contentType.StartsWith("application/x-www-form-urlencoded") &&
                        !contentType.StartsWith("application/json"))
                    {
                        LogSecurityEvent(SecurityEventType.SuspiciousInput, $"Suspicious Content-Type: {contentType}", request);
                        return false;
                    }
                }

                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error validating request source");
                return false;
            }
        }

        public string SanitizeInput(string input, int maxLength = 1000)
        {
            if (string.IsNullOrWhiteSpace(input))
                return string.Empty;

            var sanitized = input.Trim();

            // Remove HTML tags
            sanitized = Regex.Replace(sanitized, @"<[^>]*>", "", RegexOptions.IgnoreCase);

            // Remove script content
            sanitized = Regex.Replace(sanitized, @"<script[^>]*>.*?</script>", "", 
                RegexOptions.IgnoreCase | RegexOptions.Singleline);

            // Remove dangerous JavaScript events
            sanitized = Regex.Replace(sanitized, @"on\w+\s*=", "", RegexOptions.IgnoreCase);

            // Remove SQL injection patterns
            foreach (var pattern in _suspiciousPatterns)
            {
                sanitized = sanitized.Replace(pattern, "", StringComparison.OrdinalIgnoreCase);
            }

            // Remove null bytes and control characters
            sanitized = Regex.Replace(sanitized, @"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]", "");

            // Encode remaining special characters
            sanitized = System.Net.WebUtility.HtmlEncode(sanitized);

            // Truncate to max length
            if (sanitized.Length > maxLength)
            {
                sanitized = sanitized.Substring(0, maxLength);
            }

            return sanitized;
        }

        public async Task<FileSecurityValidationResult> ValidateFileSecurityAsync(IFormFile file)
        {
            var result = new FileSecurityValidationResult { IsValid = true };

            try
            {
                // Basic file validation
                if (file == null || file.Length == 0)
                {
                    result.IsValid = false;
                    result.Errors.Add("No file provided or file is empty");
                    return result;
                }

                // File size validation
                if (file.Length > _maxFileSize)
                {
                    result.IsValid = false;
                    result.Errors.Add($"File size ({file.Length} bytes) exceeds maximum allowed size ({_maxFileSize} bytes)");
                }

                if (file.Length < 1024) // Minimum 1KB
                {
                    result.IsValid = false;
                    result.Errors.Add("File is too small and may be corrupted");
                }

                // Filename validation
                if (string.IsNullOrWhiteSpace(file.FileName))
                {
                    result.IsValid = false;
                    result.Errors.Add("File must have a valid name");
                    return result;
                }

                // Check for dangerous filename patterns
                if (ContainsDangerousPatterns(file.FileName))
                {
                    result.IsValid = false;
                    result.Errors.Add("Filename contains dangerous patterns");
                    result.ThreatType = "Dangerous filename";
                }

                // Extension validation
                var extension = Path.GetExtension(file.FileName).ToLowerInvariant();
                if (_dangerousExtensions.Contains(extension))
                {
                    result.IsValid = false;
                    result.Errors.Add($"File extension '{extension}' is not allowed for security reasons");
                    result.ThreatType = "Dangerous file extension";
                }

                if (!_allowedFileExtensions.Contains(extension))
                {
                    result.IsValid = false;
                    result.Errors.Add($"File extension '{extension}' is not allowed");
                }

                // MIME type validation
                if (!_allowedMimeTypes.Contains(file.ContentType.ToLowerInvariant()))
                {
                    result.IsValid = false;
                    result.Errors.Add($"File type '{file.ContentType}' is not supported");
                }

                // File content validation (magic number check)
                await ValidateFileContentAsync(file, extension, result);

                // Check for embedded threats
                await ScanForEmbeddedThreatsAsync(file, result);

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error during file security validation");
                result.IsValid = false;
                result.Errors.Add("File validation failed due to internal error");
                return result;
            }
        }

        public string GenerateSecureToken()
        {
            using var rng = RandomNumberGenerator.Create();
            var tokenBytes = new byte[32]; // 256-bit token
            rng.GetBytes(tokenBytes);
            
            return Convert.ToBase64String(tokenBytes)
                .Replace("+", "-")
                .Replace("/", "_")
                .Replace("=", "");
        }

        public bool ValidateSecurityToken(string token)
        {
            if (string.IsNullOrWhiteSpace(token))
                return false;

            // Check token length (should be around 43 characters for 32-byte base64url)
            if (token.Length < 32 || token.Length > 50)
                return false;

            // Check for valid base64url characters
            if (!Regex.IsMatch(token, @"^[A-Za-z0-9_-]+$"))
                return false;

            return true;
        }

        public void LogSecurityEvent(SecurityEventType eventType, string details, HttpRequest request)
        {
            var clientIp = GetClientIpAddress(request);
            var userAgent = request.Headers["User-Agent"].ToString();
            var requestPath = request.Path.ToString();

            _logger.LogWarning("Security Event: {EventType} | IP: {ClientIp} | Path: {RequestPath} | Details: {Details} | UserAgent: {UserAgent}",
                eventType, clientIp, requestPath, details, userAgent);

            // In a production environment, you might want to:
            // 1. Send alerts for critical security events
            // 2. Store events in a security log database
            // 3. Integrate with SIEM systems
            // 4. Implement automatic blocking for repeated violations
        }

        #region Private Methods

        private bool ContainsDangerousPatterns(string filename)
        {
            // Check for directory traversal
            if (filename.Contains("..") || filename.Contains("\\") || filename.Contains("/"))
                return true;

            // Check for null bytes
            if (filename.Contains('\0'))
                return true;

            // Check for Windows reserved names
            var reservedNames = new[] { "CON", "PRN", "AUX", "NUL", "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9" };
            var nameWithoutExtension = Path.GetFileNameWithoutExtension(filename).ToUpperInvariant();
            if (reservedNames.Contains(nameWithoutExtension))
                return true;

            // Check for suspicious patterns
            return _suspiciousPatterns.Any(pattern => filename.Contains(pattern, StringComparison.OrdinalIgnoreCase));
        }

        private async Task ValidateFileContentAsync(IFormFile file, string extension, FileSecurityValidationResult result)
        {
            try
            {
                using var stream = file.OpenReadStream();
                var buffer = new byte[8];
                await stream.ReadAsync(buffer, 0, buffer.Length);

                var isValidFileType = extension switch
                {
                    ".jpg" or ".jpeg" => buffer[0] == 0xFF && buffer[1] == 0xD8,
                    ".png" => buffer[0] == 0x89 && buffer[1] == 0x50 && buffer[2] == 0x4E && buffer[3] == 0x47,
                    ".pdf" => buffer[0] == 0x25 && buffer[1] == 0x50 && buffer[2] == 0x44 && buffer[3] == 0x46,
                    _ => true // Allow other types for now
                };

                if (!isValidFileType)
                {
                    result.IsValid = false;
                    result.Errors.Add($"File content does not match the expected format for {extension} files");
                    result.ThreatType = "File content mismatch";
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Could not validate file content for file: {FileName}", file.FileName);
                result.Warnings.Add("Could not verify file content integrity");
            }
        }

        private async Task ScanForEmbeddedThreatsAsync(IFormFile file, FileSecurityValidationResult result)
        {
            try
            {
                using var stream = file.OpenReadStream();
                using var reader = new StreamReader(stream, Encoding.UTF8, detectEncodingFromByteOrderMarks: true);
                
                // Read first 1KB to scan for embedded scripts or malicious content
                var buffer = new char[1024];
                var bytesRead = await reader.ReadAsync(buffer, 0, buffer.Length);
                var content = new string(buffer, 0, bytesRead).ToLowerInvariant();

                // Check for embedded JavaScript
                if (content.Contains("<script") || content.Contains("javascript:") || content.Contains("vbscript:"))
                {
                    result.Warnings.Add("File may contain embedded scripts");
                    result.ThreatType = "Embedded scripts detected";
                }

                // Check for suspicious URLs
                if (content.Contains("http://") || content.Contains("https://"))
                {
                    result.Warnings.Add("File contains URLs - verify they are legitimate");
                }

                // Check for executable signatures in non-executable files
                if (file.ContentType.StartsWith("image/") && (content.Contains("mz") || content.Contains("pe")))
                {
                    result.IsValid = false;
                    result.Errors.Add("Image file contains executable content");
                    result.ThreatType = "Executable content in image";
                }
            }
            catch (Exception ex)
            {
                _logger.LogDebug(ex, "Could not scan file for embedded threats: {FileName}", file.FileName);
                // Don't fail validation if we can't scan - just log it
            }
        }

        private string GetClientIpAddress(HttpRequest request)
        {
            // Check for forwarded IP first (in case of proxy/load balancer)
            var forwardedFor = request.Headers["X-Forwarded-For"].FirstOrDefault();
            if (!string.IsNullOrEmpty(forwardedFor))
            {
                return forwardedFor.Split(',')[0].Trim();
            }

            var realIp = request.Headers["X-Real-IP"].FirstOrDefault();
            if (!string.IsNullOrEmpty(realIp))
            {
                return realIp;
            }

            return request.HttpContext.Connection.RemoteIpAddress?.ToString() ?? "Unknown";
        }

        #endregion
    }
}